### PR TITLE
Change tests & travis for correct working with Fedora 31

### PR DESCRIPTION
**Pull request by SalatskySal**
Maintainers will complete the following section:
- [x] Commit messages are descriptive enough
- [x] "Signed-off-by:" line is present in each commit
- [x] Code coverage from testing does not decrease and new code is covered
- [x] JSON/YAML configuration changes are updated in the relevant schema
- [x] Changes to metadata also update the documentation for the metadata
- [x] Pull request includes link to an osbs-docs PR for user documentation updates
- [x] New feature can be disabled from a configuration file

### CODE REVIEW

File: .travis.yml [hunk 1]
```diff
@@ -9,40 +9,36 @@ env:
   matrix:
     - ACTION=bandit
       OS=fedora
-      OS_VERSION=29
-      PYTHON_VERSION=2
-      ENGINE=docker
-    - ACTION=bandit
```
Comment (MartinBasti) [inline .travis.yml]: @athos-ribeiro I'm not sure about this, is it worth to keep bandit for py2 as well? I vote for dropping bandit on py2

File: .travis.yml [hunk 2]
```diff
@@ -9,40 +9,36 @@ env:
...
-    - OS=fedora
-      OS_VERSION=29
+      OS_VERSION=31
       PYTHON_VERSION=3
       ENGINE=docker
     - OS=fedora
       OS_VERSION=30
       PYTHON_VERSION=3
       ENGINE=docker
+    - ACTION=bandit
```
Comment (MartinBasti) [inline .travis.yml]: Please move this to original position at the top of document

File: tests/test_util.py [hunk 3]
```diff
@@ -104,13 +104,6 @@
 }
 
 
-class CustomTestException(Exception):
```
Comment (MartinBasti) [inline tests/test_util.py]: Nice!

File: tests/test_cachito_util.py [hunk 4]
```diff
@@ -63,8 +63,8 @@ def handle_request_sources(http_request):
     response = api.request_sources(CACHITO_REQUEST_REPO, CACHITO_REQUEST_REF, **additional_params)
     assert response['id'] == CACHITO_REQUEST_ID
 
-    response_json = json.dumps(response_data, indent=4)
-    assert 'Cachito response:\n{}'.format(response_json) in caplog.text
+    response_json = 'Cachito response:\n{}'.format(json.dumps(response_data, indent=4))
+    assert "".join(response_json.split()) in "".join(caplog.text.split())
```
Comment (MartinBasti) [inline tests/test_cachito_util.py]: Could be this written in more readable way?
```
>>> "something with many          whitespaces".replace(" ", "")                                                                                                                                                                            
'somethingwithmanywhitespaces'
```

Please put a comment above, why this change is needed.

**Comment by MartinBasti** [comment]
I don't think that this PR deserves 13 commits. Could you please rebase and merge commits into 1 or 2 ?

File: atomic_reactor/plugins/pre_koji_parent.py [hunk 5]
```diff
@@ -33,8 +33,8 @@
 
 class KojiParentBuildMissing(ValueError):
     """Expected to find a build for the parent image in koji, did not find it within timeout."""
-    pass
-
+    def __init__(self, error):
```
Comment (ben-alkov) [inline atomic_reactor/plugins/pre_koji_parent.py]: I don't understand how this change relates to the rest of this PR...

Comment (MartinBasti) [reply atomic_reactor/plugins/pre_koji_parent.py]: New pylint screams. However I'm not sure if this fix is correct.

why not just remove `pass` without adding `__init__`?

File: tests/test_cachito_util.py (hunk 4 above)
Comment (SalatskySal) [reply tests/test_cachito_util.py]: Do I need to put it above each changing?

Comment (MartinBasti) [reply tests/test_cachito_util.py]: IMHO Yes

Comment (ben-alkov) [reply tests/test_cachito_util.py]: IMHO, **definitely**; I didn't even realize what was going on here until I saw the comment :slightly_smiling_face:

File: tests/test_util.py [hunk 6]
```diff
@@ -58,10 +58,7 @@
                                  read_yaml, read_yaml_from_file_path, OSBSLogs,
                                  get_platforms_in_limits, get_orchestrator_platforms,
                                  dump_stacktraces, setup_introspection_signal_handler,
-                                 allow_repo_dir_in_dockerignore,
```
Comment (MartinBasti) [inline tests/test_util.py]: :eyes:

File: tests/test_cachito_util.py [hunk 7]
```diff
@@ -192,10 +196,11 @@ def handle_wait_for_request(http_request):
     expect_in_logs = dedent(
         """\
         Request {} is in "{}" state: Unknown
-        Details: {}
+         Details: {}
```
Comment (MartinBasti) [inline tests/test_cachito_util.py]: Is this change needed?

File: tests/test_util.py [hunk 8]
```diff
@@ -1645,49 +1635,3 @@ def test_allow_repo_dir_in_dockerignore(tmpdir, dockerignore_exists):
...
-    ),
-    (
-        ['com.redhat.delivery.operator.bundle=meh',
-         'com.redhat.delivery.appregistry=meh'],
-        [],
-        [has_operator_bundle_manifest, has_operator_appregistry_manifest],
-    )
-
-])
-def test_has_operator_manifest(tmpdir, workflow, labels, f_true, f_false):
```
Comment (MartinBasti) [inline tests/test_util.py]: This shouldn't be removed

**Comment by chmeliik** [comment]
Just realized, that symlink command should have been `$RUN ln -sf /usr/bin/python3 /usr/bin/python`, my bad (otherwise the command is executed on your PC instead of inside the container)

But if the current solution fixes it that would be better

File: tests/test_util.py [hunk 9]
```diff
@@ -1646,7 +1639,6 @@ def test_allow_repo_dir_in_dockerignore(tmpdir, dockerignore_exists):
         assert ignore_lines[0:len(ignore_content)] == ignore_content
         assert ignore_lines[-1] == added_lines
 
-
```
Comment (MartinBasti) [inline tests/test_util.py]: Unwanted change

File: tests/test_util.py [hunk 10]
```diff
@@ -1670,7 +1662,6 @@ def test_allow_repo_dir_in_dockerignore(tmpdir, dockerignore_exists):
         [],
         [has_operator_bundle_manifest, has_operator_appregistry_manifest],
     )
-
```
Comment (MartinBasti) [inline tests/test_util.py]: Unwanted change

**Comment by MartinBasti** [comment]
Could we just run this as to avoid such error:
```python
py_bin = "python3" if six.PY3 else "python"
subprocess.check_call([python_bin, "setup.py", "sdist", "--dist-dir", tmpdir])
```

File: tests/test_cachito_util.py (hunk 7 above)
Comment (SalatskySal) [reply tests/test_cachito_util.py]: Done

File: tests/test_util.py (hunk 9 above)
Comment (SalatskySal) [reply tests/test_util.py]: Fixed

File: tests/test_util.py (hunk 10 above)
Comment (SalatskySal) [reply tests/test_util.py]: Fixed

**Comment by SalatskySal** [comment]
Thanks you all for your help. It was important for me

File: .travis.yml [hunk 11]
```diff
@@ -14,35 +14,31 @@ env:
...
       ENGINE=docker
     - OS=fedora
       OS_VERSION=30
       PYTHON_VERSION=3
       ENGINE=docker
+    - ACTION=pylint
+      OS=fedora
+      OS_VERSION=29
+      PYTHON_VERSION=2
+      ENGINE=docker
```
Comment (chmeliik) [inline .travis.yml]: Could this be moved next to the other `pylint` action? It would be nice to have them grouped together.

File: test.sh [hunk 12]
```diff
@@ -67,6 +67,10 @@ if [[ $OS == "fedora" ]]; then
   if [[ $PYTHON_VERSION == 2* ]]; then
     $RUN $PKG $ENABLE_REPO install -y python-backports-lzma
   fi
+
+  if [[ $PYTHON_VERSION == 3* ]]; then
+    $RUN ln -sf /usr/bin/python3 /usr/bin/python
+  fi
```
Comment (chmeliik) [inline test.sh]: @SalatskySal have you tried the solution @MartinBasti suggested? That could work too, and you wouldn't need the symlink in the container.

See https://github.com/containerbuildsystem/atomic-reactor/blob/e669a655fde97e213fe72c426fec5390bb7ccce8/atomic_reactor/buildimage.py#L117

If that doesn't work, at least add a comment like `'python' command is expected to be available in container`

File: .travis.yml (hunk 11 above)
Comment (athos-ribeiro) [reply .travis.yml]: I am wondering if we really want this here. Fedora 29 Has reached its EOL in the end of 2019. IMHO, we should just stop running the CI in F29. If this means no more py2 tests in Fedora, so be it. We can keep running py2 tests in CentoOS7. What do you think?

File: .travis.yml (hunk 1 above)
Comment (athos-ribeiro) [reply .travis.yml]: Giving this a second thought,  we should drop F29 completely. Would it be possible to run bandit for py2 somewhere else? e.g., CentOS7? We could try to push the package to EPEL if it is not there :)

**Comment by athos-ribeiro** [comment]
Could you add a reference for an OSBS ticket in the commit message? See other commit messages in this project for reference.

File: tests/plugins/test_build_source_container.py [hunk 13]
```diff
@@ -187,7 +187,8 @@ def check_check_output(args, **kwargs):
         build_result = runner.run()
         err_msg = "No SRPMs directory '{}' available".format(sources_dir_path)
         err_msg += "\nNo Remote source directory '{}' available".format(remote_dir_path)
-        assert err_msg in caplog.text
+        # Since Python 3.7 logger adds additional whitespaces by default -> checking without them
+        assert err_msg.replace(" ", "") in caplog.text.replace(" ", "")
```
[the remark this answers is not in the document]
Comment (SalatskySal) [reply tests/plugins/test_build_source_container.py]: The error message remains the same, but we create new strings without white spaces and check them, so it shouldn't make misunderstandings

File: test.sh (hunk 12 above)
Comment (SalatskySal) [reply test.sh]: @chmeliik Oh, sorry, I didn`t see his solution. @MartinBasti Thanks =)

File: .travis.yml (hunk 11 above)
Comment (MartinBasti) [reply .travis.yml]: I'd keep it there until we defer py2 support. It's just linting and newer versions of pylint stopped supporting py2. So F29 is the most recent we can use for py2 linting

File: tests/plugins/test_build_source_container.py (hunk 13 above)
[the remark this answers is not in the document]
Comment (athos-ribeiro) [reply tests/plugins/test_build_source_container.py]: What I mean here is that, if the assertion fails, the error message will be harder to read for the reader.

For instance, we would see something like this:

```
AssertionError: assert 'NoSRPMsdirectory/sourcesavailable\nNoRemotesourcedirectory/remotesourcesavailable' in NoSRPMsdirectory/sourcesavailable\nNoRemotesourcedirectory/remotesourceavailable'
```

This is also error prone since it is verifying a modified version of the stings and not the actual strings as you said. We are creating new strings and checking the newly created strings instead of checking whatever output comes from the plugin being tested.

[the remark this answers is not in the document]
Comment (SalatskySal) [reply tests/plugins/test_build_source_container.py]: @athos-ribeiro You`re right, I didn`t understand you :) 
It will be better, if I change those messages by regex.

File: .travis.yml (hunk 1 above)
Comment (chmeliik) [reply .travis.yml]: Agreed, the package is installed using pip anyway so it should work for CentOS too

**Comment by MartinBasti** [comment]
no release notes needed, CI change only
